File: backend/app/schemas/system.py
"""Schemas for the system report."""

from __future__ import annotations

from pydantic import Field

from app.schemas.common import CamelModel


class TorchStatus(CamelModel):
    available: bool = Field(description="Whether PyTorch could be imported")
    version: str | None = Field(default=None, description="e.g. 2.7.1+cu118")
    cuda_version: str | None = Field(
        default=None, description="CUDA version torch was built against"
    )
    cuda_available: bool = Field(description="Whether a usable CUDA device was found")
    import_error: str | None = Field(
        default=None, description="Why torch failed to import, when it did"
    )


class GpuStatus(CamelModel):
    name: str
    vram_total_mb: int
    vram_free_mb: int = Field(description="Sampled at request time")
    capability: str = Field(description="CUDA compute capability, e.g. 8.6")


class SystemResponse(CamelModel):
    device: str = Field(description="'cuda' or 'cpu' — where inference will run")
    device_reason: str = Field(description="Why that device was selected")
    torch: TorchStatus
    gpu: GpuStatus | None = Field(
        default=None, description="Null when CUDA is unavailable, which is a normal state"
    )
    cpu_name: str
    cpu_cores_physical: int | None = None
    cpu_cores_logical: int | None = None
    ram_total_mb: int
    ram_available_mb: int
    python_version: str
    platform: str
    fp16: bool = Field(description="Whether half precision will be used")
    tile_size: int
    tile_pad: int

File: frontend/src/types/system.ts
/**
 * System and model types.
 *
 * These mirror the response schemas in `backend/app/schemas/system.py` and
 * `model.py`. The backend serialises camelCase, so no renaming happens here.
 */

export type DeviceType = 'cuda' | 'cpu'

export interface TorchStatus {
  available: boolean
  version: string | null
  cudaVersion: string | null
  cudaAvailable: boolean
  importError: string | null
}

export interface GpuStatus {
  name: string
  vramTotalMb: number
  vramFreeMb: number
  /** CUDA compute capability, e.g. "8.6". */
  capability: string
}

export interface SystemInfo {
  device: DeviceType
  /** Why that device was selected, shown verbatim in Settings. */
  deviceReason: string
  torch: TorchStatus
  /** Null when CUDA is unavailable — a normal state, not an error. */
  gpu: GpuStatus | null
  cpuName: string
  cpuCoresPhysical: number | null
  cpuCoresLogical: number | null
  ramTotalMb: number
  ramAvailableMb: number
  pythonVersion: string
  platform: string
  fp16: boolean
  tileSize: number
  tilePad: number
}

export interface ModelInfo {
  id: string
  name: string
  description: string
  /** Network architecture, e.g. "RRDBNet". */
  arch: string
  /** Native upscale factor of these weights. */
  scale: number
  supportsDenoise: boolean
  /**
   * Upscale factors this model can actually produce, published by the backend
   * so the UI never offers a combination a job would refuse.
   */
  supportedScales: number[]
  downloaded: boolean
  sizeMb: number | null
}
